### ISSUE TITLE
Link to internal rustdocs

### ISSUE THREAD

**Issue by Michael-F-Bryan** [rust-lang-nursery/rustc-guide#100]
When rust-lang/rust-central-station#40 lands we'll have a stable location which contains internal API documentation (i.e. the equivalent of `./x.py doc` with `crate-docs = true`) :tada:

This means when referring to a particular struct or function we'll be able too link to its API docs instead of linking to the original source file and hoping for the best.

Once that PR is merged we'll need to:

- [x] Update the code index to point at the API docs instead of source code
- [x] Emphasise linking to API docs where possible when people are contributing new content
- [x] Update the "contributing" section of the README appropriately (#208)
- [x] Update the *"How to build the compiler and run what you built"* to mention building docs and where they can be found online?

**Comment by rajcspsg** [comment]
@Michael-F-Bryan  - I'm new to rust. Since I see this is Easy tagged issue shall I take this issue?

**Comment by Michael-F-Bryan** [comment]
@rajcspsg if you'd like to give it a go that's fine with me, I'm also happy to mentor or answer any questions you encounter while trying to tick this issue off :grin:

We'll need to wait for rust-lang/rust-central-station#40 to be fully merged, but after that it should be relatively straightforward. If you subscribe to that issue, GitHub should send you notifications when it is resolved.

**Comment by rajcspsg** [comment]
@Michael-F-Bryan _ Thanks for mentoring me and I'm happy to take your guidance as I'm new to rust.
While we are waiting for related PR to be merged, could you please provide the file which needs to be updated with details.

**Comment by Michael-F-Bryan** [comment]
It should mainly be a case of looking for all links to source files in the `rust-lang/rust` repo and replacing them with links to the corresponding API docs.  

The easiest way would be to use a tool like grep or ripgrep and search for anything with "http" in it.

```
$ rg 'https?://'
```

**Comment by mark-i-m** [comment]
To clarify, you are looking for links in _this_ repo _to_ the other repo :)

**Comment by rajcspsg** [comment]
Thanks for the suggestion @mark-i-m @Michael-F-Bryan. 

When I searched for [here](https://doc.rust-lang.org/nightly-rustc/) for `rustc_driver::CompileState`, I end up at [stable page documentation](https://doc.rust-lang.org/1.1.0/rustc_driver/driver/struct.CompileState.html) of rustc_driver version 1.1.0.

I should have get the nightly documentation, correct?

**Comment by mark-i-m** [comment]
@rajcspsg IIUC, the new docs should be available tomorrow after the next nightly is released.

**Comment by rajcspsg** [comment]
Thanks @Michael-F-Bryan @mark-i-m.  
I've 2 questions on searching the documentation. 

1. I searched for DocContext mentioned in this [link](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/appendix-code-index.md)  using keyword `rustdoc::core` in  [DuckDuckGo](https://duckduckgo.com/) and get this [result](https://doc.rust-lang.org/1.1.0/rustdoc/core/index.html). But when I searched for `hir::Crate` I couldn't get any rust documentation. Are those going to updated with tomorrow documentation release?

2. All the links to chapter return 404 error. For example , the parser in page [appendix-code-index.md ](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/appendix-code-index.md)  returns  [url error](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/the-parser.html). Is it need to be fixed?

**Comment by rajcspsg** [comment]
@mark-i-m - Sure. I will try to fix 404 errors in github as well. I will wait for a day till document release and will let you know if I need any help.

**Comment by Michael-F-Bryan** [comment]
@rajcspsg the documentation is [now live] :tada:

[now live]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/

**Comment by rajcspsg** [comment]
Thanks @Michael-F-Bryan @mark-i-m I will work on this and submit PR.

**Comment by rajcspsg** [comment]
@Michael-F-Bryan @mark-i-m 

I've created PR for this documentation link. But I couldn't find link for below items.
Could you please help me.

| s# | FileName                  | github Link                                                                               | line# | 
|----|---------------------------|-------------------------------------------------------------------------------------------|-------| 
| 1  | const-eval.md             | https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/const_eval.rs    | 38    | 
| 2  | compiletest.md            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/runtest.rs        | 186   | 
| 3  | compiletest.md            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/header.rs         | 234   | 
| 4  | compiletest.md            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/common.rs         | 235   | 
| 5  | intro.md                  | https://github.com/rust-lang/rust/tree/master/src/bootstrap/test.rs                       | 169   | 
| 6  | miri.md                   | https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/const_eval.rs    | 115   | 
| 7  | miri.md                   | https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/step.rs          | 138   | 
| 8  | appendix-stupid-stats.md  | https://github.com/rust-lang/rust/tree/master/src/librustc_driver/lib.rs                  | 114   | 
| 9  | traits-lowering-module.md | https://github.com/rust-lang/rust/tree/master/src/test/ui/chalkify/lower_impl.rs          | 64    | 
| 10 | adding.md                 | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/header.rs         | 166   | 
| 11 | adding.md                 | https://github.com/rust-lang/rust/blob/master/src/test/ui/hello_world/main.rs             | 248   | 
| 12 | adding.md                 | https://github.com/rust-lang/rust/blob/master/src/test/ui/transmute/main.rs               | 312   | 
| 13 | type-checking.md          | https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/collect.rs              | 42    | 
| 14 | method-lookup.md          | https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/method/probe.rs   | 41    | 
| 15 | method-lookup.md          | https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/method/confirm.rs | 42    |

**Comment by mark-i-m** [comment]
Thanks! Nice job :)

> I've created PR for this documentation link. But I couldn't find link for below items.

Hmm... I can't find those either. I asked a followup on https://github.com/rust-lang/rust-central-station/pull/40

> Also, for the missing link as we disccussed here, do I need to update github link or documenation link?

Do you mean with respect to these links that you mentioned before?

> All the links to chapter return 404 error. For example , the parser in page appendix-code-index.md returns url error. Is it need to be fixed?

These should be links to [hosted version of the rustc guide](https://rust-lang-nursery.github.io/rustc-guide/) itself. I don't think we were expecting anyone to use the github interface to view the book. I'm not sure if we want to make these links absolute or leave them relative as they are... I opened #109 to deal with that; for now, let's leave them as-is.

**Comment by mark-i-m** [comment]
Also, I will try to look at the PR soon

**Comment by mark-i-m** [comment]
Ok, that was fast.

The reason those items you [listed]() are not in the hosted docs is that they are private, whereas rustdoc generates documentation for only public items by default. There is active discussion on generating docs for these private items in https://github.com/rust-lang/rust/issues/29893.

**Comment by rajcspsg** [comment]
sure Thanks @mark-i-m !!!

**Comment by rajcspsg** [comment]
@mark-i-m 
Thanks for accepting my PR :)
sure I will work on the pending items tonight and submit PR. 
I will should be able to find remaining items in same nightly documentation search as others, correct?

**Comment by rajcspsg** [comment]
Hey @mark-i-m 

I'm still not able to find documentation for below items. 

| s#| FileName                  | github Link                                                                               | line# | 
|---|---------------------------|-------------------------------------------------------------------------------------------|-------| 
| 1 | [compiletest.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/compiletest.md)            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/runtest.rs        | 186   | 
| 2 | [compiletest.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/compiletest.md)            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/header.rs         | 234   | 
| 3 | [compiletest.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/compiletest.md)            | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/common.rs         | 235   | 
| 4 | [intro.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/tests/intro.md)                  | https://github.com/rust-lang/rust/tree/master/src/bootstrap/test.rs                       | 169   | 
| 5 | [appendix-stupid-stats.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/appendix-stupid-stats.md)  | https://github.com/rust-lang/rust/tree/master/src/librustc_driver/lib.rs                  | 114   | 
| 6 | [traits-lowering-module.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/traits-lowering-module.md) | https://github.com/rust-lang/rust/tree/master/src/test/ui/chalkify/lower_impl.rs          | 64    | 
| 7 | [adding.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/tests/adding.md)                 | https://github.com/rust-lang/rust/tree/master/src/tools/compiletest/src/header.rs         | 166   | 
| 8 | [adding.md ](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/tests/adding.md)                | https://github.com/rust-lang/rust/blob/master/src/test/ui/hello_world/main.rs             | 248   | 
| 9 | [adding.md](https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/tests/adding.md)                 | https://github.com/rust-lang/rust/blob/master/src/test/ui/transmute/main.rs               | 312   | 

All of them looks like tests but I don't know how to get its documentation link.

**Comment by mark-i-m** [comment]
Thanks :)

I think anything in src/tools is not part of the compiler proper, so it will not be documented. Likewise, anything in src/test is a test, rather than part of the compiler itself, so those will not be documented either. I think we can leave these links for now.

For (5), I think we can use the following link: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/index.html

**Comment by rajcspsg** [comment]
@mark-i-m - I've updated the documentation link for https://github.com/rust-lang/rust/tree/master/src/librustc_driver/lib.rs in the PR.

**Comment by rajcspsg** [comment]
@mark-i-m @Michael-F-Bryan I think still 2 below items are yet to completed in this issue, correct?
 Update the "contributing" section of the README appropriately
 Update the "How to build the compiler and run what you built" to mention building docs and where they can be found online?

**Comment by mark-i-m** [comment]
@rajcspsg Yes, that's correct

**Comment by rajcspsg** [comment]
@mark-i-m i may need some pointers to start with. What should be updated in contributing section?

**Comment by mark-i-m** [comment]
Perhaps we can start a "Contributor Tips" section at the end? The first tip can be something like "Whenever referring to a part of the rustc source code, please add a link to the [rustc rustdocs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/). Since these links are checked, it will help us discover when something in the compiler changes and the guide needs to be updated."

**Comment by mark-i-m** [comment]
@Michael-F-Bryan did you have something specific in mind?

**Comment by Michael-F-Bryan** [comment]
Oops, looks like I missed this somehow!

> @Michael-F-Bryan did you have something specific in mind?

I was just thinking we should mention in the contributor guide that the internal API docs are available online and to link to them where possible. The comment you made sounds pretty good.

> Whenever referring to a part of the rustc source code, please add a link to the rustc rustdocs. Since these links are checked, it will help us discover when something in the compiler changes and the guide needs to be updated.

**Comment by mark-i-m** [comment]
Ping @rajcspsg Are you still interested in working on this?

**Comment by gnzlbg** [comment]
I actually have had a lot of trouble looking for the API docs of rustc, they should at least be linked in the readme, or have a section somewhere with a link to the docs. FWIW, this is the link:  https://doc.rust-lang.org/nightly/nightly-rustc/rustc/

**Comment by mark-i-m** [comment]
I agree. The docs are surprisingly hard to Google.

**Comment by nikomatsakis** [comment]
Wow, there are a lot of comments here. I recently opened up https://github.com/rust-lang-nursery/rustc-guide/pull/189, which added links to the rustdoc for the HIR section. I was thinking we should open a fresh issue (this one is perhaps too cluttered) and make it into a "quest issue". Basically, make a checklist of modules that need work, and issue out a call to try to get people involved. Seems like a nice "starter point". But maybe that's not necessary, even?

**Comment by rajcspsg** [comment]
@nikomatsakis @mark-i-m Hello guys. I'm really sorry that I'm out for a while due to few personal commitments. I think it would be good to open new issue for pending things for this PR.

**Comment by mark-i-m** [comment]
I'll try to curate such a list and open a new issue sometime soon. I'm not sure if I will have time today, though. I will post back here soon.

**Comment by rajcspsg** [comment]
@mark-i-m does this PR https://github.com/rust-lang-nursery/rustc-guide/pull/195 covers point 4 this is issue?

**Comment by mark-i-m** [comment]
Hmm... do we ever mention where they can be found? (https://doc.rust-lang.org/nightly/nightly-rustc/rustc/)

Sorry for the delay on this. I have been pretty swamped for the past few weeks (and will probably continue to be swamped until at least Oct 12)... I appreciate your patience!

**Comment by mark-i-m** [comment]
Ok, so I opened #209 to track the links @rajcspsg mentioned in https://github.com/rust-lang-nursery/rustc-guide/issues/100#issuecomment-379420709. I'm not really sure how to make progress there TBH, but I don't want to lose the list of links.

I also opened #208 to update the readme, as per the checklist in the OP. @Michael-F-Bryan @rajcspsg let me know what you think.

@rajcspsg I think the only thing left here is to update the intro section of the guide itself with a note and link to the compiler rustdocs.

**Comment by rajcspsg** [comment]
@mark-i-m shall we mark `How to build the compiler and run what you built` section done.

**Comment by mark-i-m** [comment]
Hmm... do we ever mention where they can be found? (https://doc.rust-lang.org/nightly/nightly-rustc/rustc/)

**Comment by rajcspsg** [comment]
@mark-i-m I will update and send PR for that. I have a question. 
Which file I have to update `compiler-documenting.md` or `how-to-build-and-run.md`.
Looks to me compiler documenting is correct place?

**Comment by mark-i-m** [comment]
Hooray! :tada: This is done :smile: 

Nice job @rajcspsg! This is a major improvement to the guide!

**Comment by rajcspsg** [comment]
Thanks @mark-i-m for accepting my PR :)

**Comment by lf-** [comment]
This issue is high in the Google results for "rustc internal rustdoc" and the link above doesn't work. This one does though: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/

**Comment by jyn514** [comment]
https://doc.rust-lang.org/nightly/nightly-rustc/ is the URL that will continue to work even if crates are renamed.
